cmd/tree-main.go

```diff
@@ -0,0 +1,218 @@
+/*
+ * MinIO Client (C) 2019 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/fatih/color"
+	"github.com/minio/cli"
+	"github.com/minio/mc/pkg/console"
+	"path/filepath"
+)
+
+const (
+	treeEntry     = "├── "
+	treeLastEntry = "└── "
+	treeNext      = "│"
+	treeLevel     = "   "
+)
+
+// trees files and folders.
+var treeCmd = cli.Command{
+	Name:   "tree",
+	Usage:  "list buckets and objects in a tree format",
+	Action: mainTree,
+	Before: setGlobalsFromContext,
+	Flags:  globalFlags,
+	CustomHelpTemplate: `NAME:
+  {{.HelpName}} - {{.Usage}}
+
+USAGE:
+  {{.HelpName}} [FLAGS] TARGET [TARGET ...]
+
+FLAGS:
+  {{range .VisibleFlags}}{{.}}
+  {{end}}
+EXAMPLES:
+   1. List buckets on Amazon S3 cloud storage in a tree format.
+      $ {{.HelpName}} s3
+
+   2. List all contents of mybucket on Amazon S3 cloud storage in a tree format.
+      $ {{.HelpName}} s3/mybucket/
+
+   3. List all contents of mybucket on Amazon S3 cloud storage on Microsoft Windows in a tree format.
+      $ {{.HelpName}} s3\mybucket\
+`,
+}
+
+// checkTreeSyntax - validate all the passed arguments
+func checkTreeSyntax(ctx *cli.Context) {
+	args := ctx.Args()
+	if !ctx.Args().Present() {
+		args = []string{"."}
+	}
+	for _, arg := range args {
+		if strings.TrimSpace(arg) == "" {
+			fatalIf(errInvalidArgument().Trace(args...), "Unable to validate empty argument.")
+		}
+	}
+	// extract URLs.
+	URLs := ctx.Args()
+
+	for _, url := range URLs {
+		_, _, err := url2Stat(url, false, nil)
+		if err != nil && !isURLPrefixExists(url, false) {
+			// Bucket name empty is a valid error for 'ls myminio',
+			// treat it as such.
+			if _, ok := err.ToGoError().(BucketNameEmpty); ok {
+				continue
+			}
+			fatalIf(err.Trace(url), "Unable to stat `"+url+"`.")
+		}
+	}
+}
+
+// doTree - list all entities inside a folder in a tree format.
+func doTree(url string, level int, leaf bool) error {
+	targetAlias, targetURL, _ := mustExpandAlias(url)
+	if !strings.HasSuffix(targetURL, "/") {
+		targetURL += "/"
+	}
+
+	clnt, err := newClientFromAlias(targetAlias, targetURL)
+	fatalIf(err.Trace(targetURL), "Unable to initialize target `"+targetURL+"`.")
```

Comment:
Return this error than `fatalIf()`

Reply:
The `err` is of type - `*probe.Error`. It wont be compatible with the top level calls - which listens for `Error` type